### ISSUE TITLE
Go Agent not using proxy settings from Java Configuration

### ISSUE THREAD

**Issue by cross311** [gocd/gocd#995]
My team has been using Go-Cd for our AWS resources for a number of months now and have finally gotten approval to have Go deploy our private datacenter resources.  Our datacenter is setup to only allow outbound traffic through a proxy.  This proxy is not invisible (not automatically used) it must be set via internet options or the (java control panel)[https://www.java.com/en/download/help/proxy_setup.xml].

After setting up the proxy on the machine that has the go agent installed, the installation still fails with connection timed out.  We also see no activity in the proxy logs as to it attempting to use the proxy.  We are able to access the go server via internet explorer on the machine due to it running requests properly through the outbound proxy.

Has anyone else experienced issues with the go agent not using proxy settings?  Is there a parameter that I am missing?

For information I even tried to edit the agent.cmd to pass in the proxy info when executing the registration phase.

``` sh
"%JAVA_CMD%" -Dhttp.proxyHost=%proxyhostURL% -Dhttp.proxyPort=%proxyPortNumber% -jar "%AGENT_DIR%\agent-bootstrapper.jar" %GO_SERVER% %GO_SERVER_PORT%
```

Any information or help would be much appreciated.

**Comment by arvindsv** [comment]
Can't claim to have ever tried it, but the Go Agent connects first to port 8153 of the server, using HTTP (only one request) and then, immediately connects to port 8154 using HTTPS. You might have some luck setting https.proxyHost. Can't say for sure.

**Comment by cross311** [comment]
@arvindsv Thanks for the quick reply.  We tried setting both.  However the agent launcher does not get past the original http request over 8153.

**Comment by matt-richardson** [comment]
Have you considered switching to a transparent proxy?
World of pain with old school proxies.

**Comment by cross311** [comment]
@matt-richardson Yes actually I have proposed this change.  However the folks in charge of the datacenter are not willing to change the architecture for this.

**Comment by arvindsv** [comment]
Took a quick look at this, and it turns out that Apache HTTP Client 3.1, which is used by Go internally to setup a connection between the Go Agent and the Go Server, does not honor http.proxyHost. Reference: [This](http://www.mail-archive.com/httpclient-user@jakarta.apache.org/msg02586.html).

Looks like specific code needs to be added, to make it happen. Probably somewhere around [here](https://github.com/gocd/gocd/blob/master/common/src/com/thoughtworks/go/util/HttpService.java#L148). If someone is willing to play around with it, and try adding that missing bit, I can help.

**Comment by arvindsv** [comment]
Another reference [here](http://stackoverflow.com/questions/4324089/can-i-set-the-proxy-on-the-command-line-when-using-org-apache-commons-httpclient).

**Comment by cross311** [comment]
@arvindsv, are you thinking along the lines of having a method in the httpclientfactory that takes in the httpclient and adds the proxy properties if they are in System.getProperty('')?

**Comment by arvindsv** [comment]
Oops. Didn't see. Yes, that's what I was thinking. I don't have a non-transparent proxy setup, to be able to verify it, etc.

If you want, you can make a change, use this [docker image](https://registry.hub.docker.com/u/gocd/gocd-build-installer/) to build installers, and see if it works. You'll need both agent and server installers, since the server will downgrade/upgrade agents to a compatible version (meaning, you cannot run a "modified agent" against an existing server).

**Comment by arvindsv** [comment]
@ketan: Do your recent changes for #2236 help with this?

**Comment by ketan** [comment]
Not quite, but once the PR is merged, it should be fairly straight forward to change to using proxies. I'll submit another PR after #2236 is merged.

**Comment by clearly** [comment]
Any news on this? Kinda needing it...soon.

**Comment by ketan** [comment]
For anyone willing to provide feedback on an experimental build, please see my note at https://github.com/gocd/gocd/pull/3847#issuecomment-329248430
